File: BackEnd/app/schemas/listing.py
from pydantic import BaseModel, ConfigDict, model_validator
from datetime import datetime
from typing import Optional, List
from app.models.enums import ListingStatusEnum, ModerationStatusEnum, PromotionStatusEnum

class ListingImageResponse(BaseModel):
    id: int
    file_url: str
    is_primary: bool
    order_index: int
    model_config = ConfigDict(from_attributes=True)

class ListingCreate(BaseModel):
    title: str
    description: str
    price: float
    currency: str = "USD"
    city: str
    category_id: int
    is_negotiable: bool = False
    # OSM / map picker: decimal degrees (WGS84). Omit both if no pin yet.
    latitude: Optional[float] = None
    longitude: Optional[float] = None

    @model_validator(mode="after")
    def coordinates_both_or_neither(self) -> "ListingCreate":
        lat, lon = self.latitude, self.longitude
        if (lat is None) != (lon is None):
            raise ValueError("latitude and longitude must be provided together, or both omitted")
        if lat is not None:
            if lat < -90 or lat > 90:
                raise ValueError("latitude must be between -90 and 90")
            if lon is not None and (lon < -180 or lon > 180):
                raise ValueError("longitude must be between -180 and 180")
        return self


class ListingUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = None
    currency: Optional[str] = None
    city: Optional[str] = None
    category_id: Optional[int] = None
    is_negotiable: Optional[bool] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None

    @model_validator(mode="after")
    def coordinates_both_or_neither(self) -> "ListingUpdate":
        lat, lon = self.latitude, self.longitude
        if (lat is None) != (lon is None):
            raise ValueError("latitude and longitude must be provided together, or both omitted")
        if lat is not None:
            if lat < -90 or lat > 90:
                raise ValueError("latitude must be between -90 and 90")
            if lon is not None and (lon < -180 or lon > 180):
                raise ValueError("longitude must be between -180 and 180")
        return self

class ListingResponse(ListingCreate):
    id: int
    owner_id: int
    status: ListingStatusEnum
    moderation_status: ModerationStatusEnum
    promotion_status: PromotionStatusEnum
    view_count: int
    created_at: datetime
    images: List[ListingImageResponse] = []
    
    model_config = ConfigDict(from_attributes=True)

File: Mobile/lib/features/listings/data/models/listing.dart
import 'package:json_annotation/json_annotation.dart';

part 'listing.g.dart';

@JsonSerializable()
class ListingImage {
  final int id;
  
  @JsonKey(name: 'file_url')
  final String fileUrl;
  
  @JsonKey(name: 'is_primary')
  final bool isPrimary;
  
  @JsonKey(name: 'order_index')
  final int orderIndex;

  ListingImage({
    required this.id,
    required this.fileUrl,
    required this.isPrimary,
    required this.orderIndex,
  });

  factory ListingImage.fromJson(Map<String, dynamic> json) =>
      _$ListingImageFromJson(json);

  Map<String, dynamic> toJson() => _$ListingImageToJson(this);
}

@JsonSerializable()
class Listing {
  final int id;
  
  @JsonKey(name: 'owner_id')
  final int ownerId;
  
  final String title;
  final String description;
  final double price;
  final String currency;
  final String city;

  /// WGS84 decimal degrees from map/OSM; both null if not pinned.
  final double? latitude;
  final double? longitude;
  
  @JsonKey(name: 'category_id')
  final int categoryId;
  
  @JsonKey(name: 'is_negotiable')
  final bool isNegotiable;
  
  final String status;
  
  @JsonKey(name: 'moderation_status')
  final String moderationStatus;
  
  @JsonKey(name: 'promotion_status')
  final String promotionStatus;
  
  @JsonKey(name: 'view_count')
  final int viewCount;
  
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  
  final List<ListingImage> images;

  Listing({
    required this.id,
    required this.ownerId,
    required this.title,
    required this.description,
    required this.price,
    required this.currency,
    required this.city,
    this.latitude,
    this.longitude,
    required this.categoryId,
    required this.isNegotiable,
    required this.status,
    required this.moderationStatus,
    required this.promotionStatus,
    required this.viewCount,
    required this.createdAt,
    this.images = const [],
  });

  factory Listing.fromJson(Map<String, dynamic> json) =>
      _$ListingFromJson(json);

  Map<String, dynamic> toJson() => _$ListingToJson(this);
}
